pybamm/discretisations/base_discretisation.py

```diff
@@ -0,0 +1,306 @@
+
+    def process_rhs(self, rhs, boundary_conditions, y_slices):
+        """Discretise initial conditions.
+
+        Parameters
+        ----------
+        rhs : dict
+            Equations ({variable: equation} dict) to dicretise
+        boundary_conditions : dict
+            Boundary conditions ({symbol.id: (left bc, right bc)} dict) associated with
+            rhs to dicretise
+        y_slices : dict of {variable id: slice}
+            The slices to assign to StateVectors when discretising a variable
+
+        Returns
+        -------
+        dydt : method
+            A method that takes a time t and array y and returns the concatenated
+            time-derivatives.
+
+        """
+        boundary_conditions = {
+            key.id: value for key, value in boundary_conditions.items()
+        }
+        for variable, equation in rhs.items():
+            rhs[variable] = self.process_symbol(
+                equation, variable.domain, y_slices, boundary_conditions
+            )
+
+        # Concatenate and evaluate right-hand sides
+        self._concatenated_rhs = pybamm.NumpyConcatenation(*rhs.values())
+
+        def dydt(t, y):
+            return self._concatenated_rhs.evaluate(t, y)
+
+        return dydt
+
+    def process_symbol(self, symbol, domain, y_slices=None, boundary_conditions={}):
+        """Discretise operators in model equations.
+
+        Parameters
+        ----------
+        symbol : :class:`pybamm.expression_tree.symbol.Symbol` (or subclass) instance
+            Symbol to discretise
+        y_slices : dict of {Variable: slice}
+            The slices to assign to StateVectors when discretising a variable
+
+        Returns
+        -------
+        :class:`pybamm.expression_tree.symbol.Symbol` (or subclass) instance
+            Discretised symbol
+
+        """
+        if isinstance(symbol, pybamm.Gradient):
+            return self.gradient(
+                symbol.children[0], domain, y_slices, boundary_conditions
+            )
+
+        if isinstance(symbol, pybamm.Divergence):
+            return self.divergence(
+                symbol.children[0], domain, y_slices, boundary_conditions
+            )
+
+        elif isinstance(symbol, pybamm.BinaryOperator):
+            new_left = self.process_symbol(
+                symbol.children[0], domain, y_slices, boundary_conditions
+            )
+            new_right = self.process_symbol(
+                symbol.children[1], domain, y_slices, boundary_conditions
+            )
+            return symbol.__class__(new_left, new_right)
+
+        elif isinstance(symbol, pybamm.UnaryOperator):
+            new_child = self.process_symbol(
+                symbol.children[0], domain, y_slices, boundary_conditions
+            )
+            return symbol.__class__(new_child)
+
+        elif isinstance(symbol, pybamm.Variable):
+            return pybamm.StateVector(y_slices[symbol.id])
+
+        elif isinstance(symbol, pybamm.Scalar):
```

Comment:
Sure

Reply:
maybe that is because the new node has its parent defined to *something*, even though it should be `None`? You are building the tree from the bottom-up, so none of the new nodes should have parents I think.....

Reply:
This is the only way I've found to do it, a bit hacky but seems to work:
```python3
parent = symbol.parent
symbol.parent = None
new_symbol = copy.copy(symbol)
symbol.parent = parent
return new_symbol
```

Reply:
It seems having to copy children to avoid loops is inevitable with a binary tree structure. The tests after that are quite stringent but will probably help us avoid trouble down the line?

Reply:
yes, that is what I hope!